### PR TITLE
Added the type registry in advance of implementing Any support.

### PR DESCRIPTION
**Pull request by jskeet**
Biting off just this bit first as I don't need the changes from a previous PR for this part.

### CODE REVIEW

**Comment by jskeet** [comment]
Not sure why mcs is complaining about the location of the XML comment. Will look into it...

**Comment by jskeet** [comment]
Moving the location of the read-only static autoprop (with initializer) seems to have fixed it. Will see if this is still a bug in latest mcs and file an issue if so.

File: csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            Preconditions.CheckNotNull(fileDescriptors, nameof(fileDescriptors));
+            var builder = new Builder();
+            foreach (var file in fileDescriptors)
+            {
+                builder.AddFile(file);
+            }
+            return builder.Build();
+        }
+
+        private static void AddMessage(Dictionary<string, MessageDescriptor> dictionary, MessageDescriptor messageDescriptor)
```
Comment (jtattermusch) [inline csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs]: Impl missing?

File: csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+#endregion
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Protobuf.Reflection
+{
+    /// <summary>
+    /// An immutable registry of types which can be looked up by their full name.
+    /// </summary>
+    public sealed class TypeRegistry
```
Comment (jtattermusch) [inline csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs]: For gRPC server reflection functionality, we will probably need to implement equivalents for FindFileByName  and FindFileContainingSymbol from C++ DescriptorDatabase. Does that sound reasonable?

https://github.com/google/protobuf/blob/24fe6ba750bf8a247d41fc589460dce1bdd697a1/src/google/protobuf/descriptor_database.h

Btw, what is supposed to be the difference between TypeRegistry and DescriptorPool?

File: csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        }
+
+        /// <summary>
+        /// Attempts to find a message descriptor by its full name.
+        /// </summary>
+        /// <param name="fullName">The full name of the message, which is the dot-separated
+        /// combination of package, containing messages and message name</param>
+        /// <returns>The message descriptor corresponding to <paramref name="fullName"/> or null
+        /// if there is no such message descriptor.</returns>
+        public MessageDescriptor Find(string fullName)
```
Comment (jtattermusch) [inline csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs]: It looks like we might be adding more Find* methods to this class in the future, so perhaps FindMessage or FindMessageByName would be better?

**Comment by jtattermusch** [comment]
It's still a bit unclear to me what exact functionality will be needed for gRPC server reflection (and what is the distinction between TypeRegistry and DescriptorPool), but I don't have anything against the code in this PR, so I think I am fine with merging (and we can figure out server reflection later).

Btw, I'll be OOO next week.

File: csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs (hunk 2 above)
Comment (jskeet) [inline csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs]: DescriptorPool is internal and contains *all* descriptors (not just messages) in a particular tree for referential consistency. TypeRegistry is just for lookup of *message* descriptors by name, for Any support.

File: csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs (hunk 1 above)
Comment (jskeet) [inline csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs]: Gah - that method shouldn't be there. It started off there before I decided to use a builder. Will delete (and squash commit).

File: csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs (hunk 3 above)
Comment (jskeet) [inline csharp/src/Google.Protobuf/Reflection/TypeRegistry.cs]: Would rather not do that speculatively just yet. This is what is required for Any support, and it's what the Java code has. But we can rename it if GRPC needs more later.

**Comment by jskeet** [comment]
Thanks - with this and #990 in, I should be able to get Any support implemented by the time you're back... we can take it from there.

**Pull request merged**